resources/type-system/inference.md

```diff
@@ -1210,6 +1247,674 @@ of steps:
+
+  - If `F` is `∅`, then:
+
+    - Let `C` be an empty list of type constraints, and let `{U_1, U_2, ...}` be
+      the same as `{T_1, T_2, ...}`. _This covers the case of dynamic
+      invocation; the user-supplied type arguments are accepted without
+      modification._
+
+  - Otherwise, `F` must be a function type. If the number of formal type
+    parameters of `F` (which could be zero) matches the number of type arguments
+    `T`, then:
+
+    - Let `C` be an empty list of type constraints, and let `{U_1, U_2, ...}` be
+      the same as `{T_1, T_2, ...}`. _This covers the case where no generic type
+      inference is needed, because a correct number of explicit type arguments
+      was supplied._
+
+  - Otherwise, if the number of type arguments `T` is nonzero, then there is a
+    compile-time error (_wrong number of type arguments supplied_).
+
+  - Otherwise, `F` must have at least one formal type parameter, and there must
+    be zero type arguments `T` (_in other words, type arguments will be
+    inferred_). Then:
+
+    - Initialize `C` to an empty list of type constraints.
+
+    - Initialize `{U_1, U_2, ...}` to a list of type schemas, of the same length
+      as `{X_1, X_2, ...}`, each of which is `_`.
+
+    - Using subtype constraint generation, attempt to match `R_F` as a subtype
+      of `K` with respect to the list of type variables `{X_1, X_2, ...}`.
+
+      - If this succeeds, then accumulate the resulting list of type constraints
+        into `C`. Then, let `{V_1, V_2, ...}` be the constraint solution for the
+        set of type variables `{X_1, X_2, ...}` with respect to the constaint
+        set `C`, with partial solution `{U_1, U_2, ...}`. Replace `{U_1, U_2,
+        ...}` with `{V_1, V_2, ...}`. _This step is known as "downwards
+        inference", since it has the effect of passing type information __down__
+        the syntax tree from the context of the invocation to the context of the
+        arguments._
+
+      - Otherwise, leave `C` and `{U_1, U_2, ...}` unchanged.
+
+- Partition the arguments `{e_1, e_2, ...}` into stages (see [argument
+  partitioning](#Argument-partitioning) below), and then for each stage _k_:
+
+  - For each `e_i` in stage _k_, let `K_i` be the result of substituting `{U_1,
+    U_2, ...}` for `{X_1, X_2, ...}` in `P_i`.
+
+  - For each `e_i` in stage _k_ that _does not_ take the form of a
+    _&lt;functionExpression&gt;_ enclosed in zero or more parentheses, in order
+    of increasing _i_:
+
+    - Let `m_i_preliminary` be the result of performing expression inference on
+      `e_i`, in context `K_i`.
+
+  - For each `e_i` in stage _k_ that _does_ take the form of a
+    _&lt;functionExpression&gt;_ enclosed in zero or more parentheses, in order
+    of increasing _i_:
+
+    - Let `m_i_preliminary` be the result of performing expression inference on
+      `e_i`, in context `K_i`.
+
+  - _Note that a property of argument partitioning is that arguments that are
```

Comment:
This raises the obvious question "why bother to order things within a stage then"?  That is, since no incremental solutions are generated *within* a phase, why traverse the non function expressions in stage 0 before the function expressions?  I think (but am not sure) the answer is just that "this is what the implementation does" combined with the fact that I *think* (but am not sure) that our constraint generation algorithm is not associative/commutative.  That is, I think (but am not sure) that doing constraint solving in a different order on a list of arguments can result in a different answer (because I think our upper bound algorithm may not be associative).  This is yet another meta-theoretic detail that would be nice to nail down for certain.  In any case, if my understanding is correct, perhaps worth noting this in the commentary, since other readers may ask the same question?

Reply:
Good question! This was actually a deliberate choice I made as part of the design of horizontal inference, because it allows any type promotions in the non-function expressions to carry over into the function expressions, regardless of the order of the arguments. There’s a detailed explanation here: https://github.com/dart-lang/language/blob/main/accepted/2.18/horizontal-inference/feature-specification.md#why-do-we-defer-all-function-literals. I’ve added a smaller explanation to this PR, with a link to the more detailed explanation.

Reply:
ok, thanks.